<?xml version="1.0" ?>
<robot name="service_robot" xmlns:xacro="http://ros.org/wiki/xacro">
  <xacro:include filename="$(find service_robot)/urdf/common_properties.xacro"/>
  <xacro:include filename="$(find service_robot)/urdf/service_robot.gazebo.xacro"/>

  <xacro:property name="r200_cam_rgb_px" value="0.005"/>
  <xacro:property name="r200_cam_rgb_py" value="0.018"/>
  <xacro:property name="r200_cam_rgb_pz" value="0.013"/>
  <xacro:property name="r200_cam_depth_offset" value="0.01"/>

  <link name="base_footprint"/>

  <joint name="base_joint" type="fixed">
    <parent link="base_footprint"/>
    <child link="base_link" />
    <origin xyz="0 0 0.010" rpy="0 0 0"/>
  </joint>

  <link name="base_link">
 <!--   <visual>
      <origin xyz="-0.064 0 0.0" rpy="0 0 0"/>
      <geometry>
        <box size="0.8 0.6 0.094"/>
      </geometry>
      <material name="light_black"/>
    </visual> -->

    <collision>
      <origin xyz="0 0 0.047" rpy="0 0 0"/>
      <geometry>
        <box size="0.7 0.5 0.094"/>
      </geometry>
    </collision>
    
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <mass value="1.3729096e+00"/>
      <inertia ixx="8.7002718e-03" ixy="-4.7576583e-05" ixz="1.1160499e-04"
               iyy="8.6195418e-03" iyz="-3.5422299e-06"
               izz="1.4612727e-02" />
    </inertial>
  </link>
  <link name="rack_0">
    <visual>
      <origin xyz="0 0 0.0" rpy="0 0 0"/>
      <geometry>
        <box size="0.7 0.5 0.294"/>
      </geometry>
      <material name="white"/>
    </visual>

  </link>
  <joint name="rack_0_joint" type="fixed">
    <parent link="base_link"/>
    <child link="rack_0" />
    <origin xyz="0 0 0.2" rpy="0 0 0"/>
  </joint>
  <link name="rack_1">
    <visual>
      <origin xyz="0 0 0.0" rpy="0 0 0"/>
      <geometry>
        <box size="0.7 0.5 0.044"/>
      </geometry>
      <material name="light_black"/>
    </visual>

  </link>
  <joint name="rack_1_joint" type="fixed">
    <parent link="base_link"/>
    <child link="rack_1" />
    <origin xyz="0 0 0.4" rpy="0 0 0"/>
  </joint>
  <link name="rack_2">
    <visual>
      <origin xyz="0 0 0.0" rpy="0 0 0"/>
      <geometry>
        <box size="0.7 0.5 0.044"/>
      </geometry>
      <material name="light_black"/>
    </visual>

  </link>
  <joint name="rack_2_joint" type="fixed">
    <parent link="base_link"/>
    <child link="rack_2" />
    <origin xyz="0 0 0.6" rpy="0 0 0"/>
  </joint>

  <link name="rack_3">
    <visual>
      <origin xyz="0 0 0.0" rpy="0 0 0"/>
      <geometry>
        <box size="0.7 0.5 0.044"/>
      </geometry>
      <material name="light_black"/>
    </visual>

  </link>
  <joint name="rack_3_joint" type="fixed">
    <parent link="base_link"/>
    <child link="rack_3" />
    <origin xyz="0 0 0.8" rpy="0 0 0"/>
  </joint>
  <link name="body">
    <visual>
      <origin xyz="-0.064 0 0.0" rpy="0 0 0"/>
      <geometry>
        <box size="0.12 0.3 0.4"/>
      </geometry>
      <material name="grey"/>
    </visual>

  </link>
  <joint name="body_joint" type="fixed">
    <parent link="base_link"/>
    <child link="body" />
    <origin xyz="-0.23 0 0.85" rpy="0 0 0"/>
  </joint>
  
  <link name="neck">
    <visual>
      <origin xyz="-0.064 0 0.0" rpy="0 0 0"/>
      <geometry>
        <box size="0.02 0.08 0.04"/>
      </geometry>
      <material name="dark"/>
    </visual>

  </link>
  <joint name="neck_joint" type="fixed">
    <parent link="base_link"/>
    <child link="neck" />
    <origin xyz="-0.23 0 1.05" rpy="0 0 0"/>
  </joint>

  <link name="head">
    <visual>
      <origin xyz="-0.064 0 0.0" rpy="0 0 0"/>
      <geometry>
        <box size="0.04 0.20 0.10"/>
      </geometry>
      <material name="black"/>
    </visual>

  </link>
  <joint name="head_joint" type="fixed">
    <parent link="base_link"/>
    <child link="head" />
    <origin xyz="-0.23 0 1.123" rpy="0 0 0"/>
  </joint>

  <joint name="wheel_left_joint" type="continuous">
    <parent link="base_link"/>
    <child link="wheel_left_link"/>
    <origin xyz="0.0 0.165 0.09" rpy="-1.57 0 0"/>
    <axis xyz="0 0 1"/>
  </joint>
 <!-- <link name="tab_holder">
    <visual>
      <origin xyz="-0.064 0 0.0" rpy="0 0 0"/>
      <geometry>
        <box size="0.02 0.15 0.15"/>
      </geometry>
      <material name="black"/>
    </visual>

  </link>
  <joint name="tab_holder_joint" type="fixed">
    <parent link="base_link"/>
    <child link="tab_holder" />
    <origin xyz="0 0 0.9" rpy="0 0 0"/>
  </joint> -->
<link name="pole_1">
     
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>


          <box size="0.02 0.02 0.4"/>


      </geometry>
      <material name="white" />
    </visual>  
  </link>

  <joint name="pole_1_joint" type="fixed">
    <origin xyz="0.3 0.25 0.6" rpy="0 0 0" />
    <parent link="base_link"/>
    <child link="pole_1" />
  </joint>
<link name="pole_2">
     
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>


          <box size="0.02 0.02 0.4"/>


      </geometry>
      <material name="white" />
    </visual>  
   
  </link>

  <joint name="pole_2_joint" type="fixed">
    <origin xyz="0.3 -0.25 0.6" rpy="0 0 0" />
    <parent link="base_link"/>
    <child link="pole_2" />
  </joint>
<link name="pole_3">
   
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>


          <box size="0.02 0.02 0.4"/>


      </geometry>
      <material name="white" />
    </visual>  
   
  </link>

  <joint name="pole_3_joint" type="fixed">
    <origin xyz="-0.35 -0.25 0.6" rpy="0 0 0" />
    <parent link="base_link"/>
    <child link="pole_3" />
  </joint>
<link name="pole_4">
    
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>


          <box size="0.02 0.02 0.4"/>


      </geometry>
      <material name="white" />
    </visual>  
     
  </link>

  <joint name="pole_4_joint" type="fixed">
    <origin xyz="-0.35 0.25 0.6" rpy="0 0 0" />
    <parent link="base_link"/>
    <child link="pole_4" />
  </joint>
  <link name="wheel_left_link">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder length="0.018" radius="0.1"/>
      </geometry>
      <material name="black"/>
    </visual>

    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder length="0.018" radius="0.1"/>
      </geometry>
    </collision>

    <inertial>
      <origin xyz="0 0 0" />
      <mass value="2.8498940e-02" />
      <inertia ixx="1.1175580e-05" ixy="-4.2369783e-11" ixz="-5.9381719e-09"
               iyy="1.1192413e-05" iyz="-1.4400107e-11"
               izz="2.0712558e-05" />
      </inertial>
  </link>

  <joint name="wheel_right_joint" type="continuous">
    <parent link="base_link"/>
    <child link="wheel_right_link"/>
    <origin xyz="0.0 -0.165 0.09" rpy="-1.57 0 0"/>
    <axis xyz="0 0 1"/>
  </joint>

  <link name="wheel_right_link">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder length="0.018" radius="0.1"/>
      </geometry>
       <material name="black"/>
    </visual>

    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder length="0.018" radius="0.1"/>
      </geometry>
    </collision>

    <inertial>
      <origin xyz="0 0 0" />
      <mass value="2.8498940e-02" />
      <inertia ixx="1.1175580e-05" ixy="-4.2369783e-11" ixz="-5.9381719e-09"
               iyy="1.1192413e-05" iyz="-1.4400107e-11"
               izz="2.0712558e-05" />
      </inertial>
  </link>

  <joint name="caster_back_right_joint" type="fixed">
    <parent link="base_link"/>
    <child link="caster_back_right_link"/>
    <origin xyz="-0.177 -0.064 -0.004" rpy="-1.57 0 0"/>
  </joint>

  <link name="caster_back_right_link">
    <collision>
      <origin xyz="0 0.001 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.030 0.009 0.020"/>
      </geometry>
    </collision>

    <inertial>
      <origin xyz="0 0 0" />
      <mass value="0.005" />
      <inertia ixx="0.001" ixy="0.0" ixz="0.0"
               iyy="0.001" iyz="0.0"
               izz="0.001" />
    </inertial>
  </link>

  <joint name="caster_back_left_joint" type="fixed">
    <parent link="base_link"/>
    <child link="caster_back_left_link"/>
    <origin xyz="-0.177 0.064 -0.004" rpy="-1.57 0 0"/>
  </joint>

  <link name="caster_back_left_link">
    <collision>
      <origin xyz="0 0.001 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.030 0.009 0.020"/>
      </geometry>
    </collision>

    <inertial>
      <origin xyz="0 0 0" />
      <mass value="0.005" />
      <inertia ixx="0.001" ixy="0.0" ixz="0.0"
               iyy="0.001" iyz="0.0"
               izz="0.001" />
    </inertial>
  </link>

  <joint name="imu_joint" type="fixed">
    <parent link="base_link"/>
    <child link="imu_link"/>
    <origin xyz="0.0 0 0.068" rpy="0 0 0"/>
  </joint>

  <link name="imu_link"/>

  <joint name="scan_joint" type="fixed">
    <parent link="base_link"/>
    <child link="base_scan"/>
    <origin xyz="-0.278 0 0.44" rpy="0 0 0"/>
  </joint>

  <link name="base_scan">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder length="0.0315" radius="0.055"/>
      </geometry>
      <material name="dark"/>
    </visual>

    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder length="0.0315" radius="0.055"/>
      </geometry>
    </collision>

    <inertial>
      <mass value="0.114" />
      <origin xyz="0 0 0" />
      <inertia ixx="0.001" ixy="0.0" ixz="0.0"
               iyy="0.001" iyz="0.0"
               izz="0.001" />
    </inertial>
  </link>

  <joint name="camera_joint" type="fixed">
    <origin xyz="0.34 0 0.41" rpy="0 0 0"/>
    <parent link="base_link"/>
    <child link="camera_link"/>
  </joint>

  <link name="camera_link">
    <visual>
     <origin xyz="0.09 0 0.2" rpy="1.57 0 1.57"/>
      <geometry>
       <box size="0.162 0.032 0.030"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.09 0 0.2" rpy="0 0 0"/>
      <geometry>
        <box size="0.162 0.032 0.030"/>
      </geometry>
    </collision>

    <!-- This inertial field needs doesn't contain reliable data!! -->
<!--   <inertial>
      <mass value="0.564" />
      <origin xyz="0 0 0" />
      <inertia ixx="0.003881243" ixy="0.0" ixz="0.0"
               iyy="0.000498940" iyz="0.0"
               izz="0.003879257" />
    </inertial>-->
  </link>

  <joint name="camera_rgb_joint" type="fixed">
    <origin xyz="${r200_cam_rgb_px} ${r200_cam_rgb_py} ${r200_cam_rgb_pz}" rpy="0 0 0"/>
    <parent link="camera_link"/>
    <child link="camera_rgb_frame"/>
  </joint>
  <link name="camera_rgb_frame"/>

  <joint name="camera_rgb_optical_joint" type="fixed">
    <origin xyz="0 0 0" rpy="-1.57 0 -1.57"/>
    <parent link="camera_rgb_frame"/>
    <child link="camera_rgb_optical_frame"/>
  </joint>
  <link name="camera_rgb_optical_frame"/>

  <joint name="camera_depth_joint" type="fixed">
    <origin xyz="${r200_cam_rgb_px} ${r200_cam_rgb_py + r200_cam_depth_offset} ${r200_cam_rgb_pz}" rpy="0 0 0"/>
    <parent link="camera_link"/>
    <child link="camera_depth_frame"/>
  </joint>
  <link name="camera_depth_frame"/>

  <joint name="camera_depth_optical_joint" type="fixed">
    <origin xyz="0 0 0" rpy="-1.57 0 -1.57"/>
    <parent link="camera_depth_frame"/>
    <child link="camera_depth_optical_frame"/>
  </joint>
  <link name="camera_depth_optical_frame"/>

</robot>
